File: pkg/client/archive_test.go
package client

import (
	"archive/tar"
	"archive/zip"
	"bytes"
	"compress/gzip"
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"sort"
	"strings"
	"sync"
	"testing"
)

// archiveMockServer is a minimal httptest harness that serves a remote tree
// (listings + inline file bodies) for ArchiveDir tests.
type archiveMockServer struct {
	mu          sync.Mutex
	listEntries map[string][]FileInfo
	fileBodies  map[string][]byte
	listCalls   map[string]int // path -> number of ListCtx calls
}

func newArchiveMockServer(files map[string]string) *archiveMockServer {
	m := &archiveMockServer{
		listEntries: map[string][]FileInfo{},
		fileBodies:  map[string][]byte{},
		listCalls:   map[string]int{},
	}
	dirSeen := map[string]map[string]bool{}
	addChild := func(dir, name string, info FileInfo) {
		if dirSeen[dir] == nil {
			dirSeen[dir] = map[string]bool{}
		}
		if dirSeen[dir][name] {
			return
		}
		dirSeen[dir][name] = true
		m.listEntries[dir] = append(m.listEntries[dir], info)
	}
	for p, body := range files {
		m.fileBodies[p] = []byte(body)
		dir := "/"
		rel := strings.TrimPrefix(p, "/")
		segs := strings.Split(rel, "/")
		for i, seg := range segs {
			isLast := i == len(segs)-1
			if isLast {
				addChild(dir, seg, FileInfo{Name: seg, Size: int64(len(body)), IsDir: false, HasMode: true, Mode: 0o644})
				break
			}
			var child string
			if dir == "/" {
				child = "/" + seg
			} else {
				child = dir + "/" + seg
			}
			addChild(dir, seg, FileInfo{Name: seg, IsDir: true, HasMode: true, Mode: 0o755})
			if m.listEntries[child] == nil {
				m.listEntries[child] = []FileInfo{}
			}
			dir = child
		}
	}
	return m
}

func (m *archiveMockServer) server(t *testing.T) *httptest.Server {
	t.Helper()
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		m.mu.Lock()
		defer m.mu.Unlock()
		switch {
		case r.Method == http.MethodGet && strings.HasPrefix(r.URL.Path, "/v1/fs/") && r.URL.Query().Has("list"):
			p := strings.TrimPrefix(r.URL.Path, "/v1/fs")
			m.listCalls[p]++
			entries := m.listEntries[p]
			out := make([]map[string]any, len(entries))
			for i, e := range entries {
				out[i] = map[string]any{
					"name": e.Name, "size": e.Size, "isDir": e.IsDir,
					"mode": e.Mode, "hasMode": e.HasMode,
				}
			}
			_ = json.NewEncoder(w).Encode(map[string]any{"entries": out})
		case r.Method == http.MethodGet && strings.HasPrefix(r.URL.Path, "/v1/fs/"):
			p := strings.TrimPrefix(r.URL.Path, "/v1/fs")
			if body, ok := m.fileBodies[p]; ok {
				_, _ = w.Write(body)
				return
			}
			http.NotFound(w, r)
		default:
			http.NotFound(w, r)
		}
	}))
}

func TestArchiveDirTarGzBasicTree(t *testing.T) {
	mock := newArchiveMockServer(map[string]string{
		"/proj/README.md":        "hello world\n",
		"/proj/src/main.go":      "package main\n",
		"/proj/src/util/util.go": "package util\n",
	})
	srv := mock.server(t)
	defer srv.Close()
	c := New(srv.URL, "")
	c.smallFileThreshold = DefaultSmallFileThreshold

	var buf bytes.Buffer
	if err := c.ArchiveDir(context.Background(), "/proj", &buf, ArchiveOptions{}); err != nil {
		t.Fatalf("ArchiveDir: %v", err)
	}
	gz, err := gzip.NewReader(bytes.NewReader(buf.Bytes()))
	if err != nil {
		t.Fatalf("gzip reader: %v", err)
	}
	defer func() { _ = gz.Close() }()
	tr := tar.NewReader(gz)
	var names []string
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			t.Fatalf("tar next: %v", err)
		}
		name := hdr.Name
		if hdr.Typeflag == tar.TypeDir {
			name = strings.TrimSuffix(name, "/") + "/"
		}
		names = append(names, name)
	}
	sort.Strings(names)
	for _, want := range []string{"proj/", "proj/README.md", "proj/src/", "proj/src/main.go", "proj/src/util/", "proj/src/util/util.go"} {
		if !containsStr(names, want) {
			t.Fatalf("archive missing %q; got %v", want, names)
		}
	}
}

func TestArchiveDirExcludeSkipsNodeModules(t *testing.T) {
	mock := newArchiveMockServer(map[string]string{
		"/proj/src/app.go":              "package src\n",
		"/proj/node_modules/react/x.js": "module.exports\n",
	})
	srv := mock.server(t)
	defer srv.Close()
	c := New(srv.URL, "")
	c.smallFileThreshold = DefaultSmallFileThreshold

	var buf bytes.Buffer
	if err := c.ArchiveDir(context.Background(), "/proj", &buf, ArchiveOptions{
		Exclude: []string{"**/node_modules/**"},
	}); err != nil {
		t.Fatalf("ArchiveDir: %v", err)
	}
	gz, _ := gzip.NewReader(bytes.NewReader(buf.Bytes()))
	defer func() { _ = gz.Close() }()
	tr := tar.NewReader(gz)
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			t.Fatalf("tar next: %v", err)
		}
		if strings.Contains(hdr.Name, "node_modules") {
			t.Fatalf("excluded node_modules leaked into archive: %q", hdr.Name)
		}
		if hdr.Name == "proj/src/app.go" {
			return
		}
	}
	t.Fatal("app.go missing from archive")
}

func TestArchiveDirIncludeWhitelist(t *testing.T) {
	mock := newArchiveMockServer(map[string]string{
		"/proj/src/app.go":    "package src\n",
		"/proj/README.md":     "# readme\n",
		"/proj/docs/guide.md": "# guide\n",
	})
	srv := mock.server(t)
	defer srv.Close()
	c := New(srv.URL, "")
	c.smallFileThreshold = DefaultSmallFileThreshold

	var buf bytes.Buffer
	if err := c.ArchiveDir(context.Background(), "/proj", &buf, ArchiveOptions{
		Include: []string{"docs/**", "README.md"},
	}); err != nil {
		t.Fatalf("ArchiveDir: %v", err)
	}
	gz, _ := gzip.NewReader(bytes.NewReader(buf.Bytes()))
	defer func() { _ = gz.Close() }()
	tr := tar.NewReader(gz)
	var foundGuide, foundReadme bool
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			t.Fatalf("tar next: %v", err)
		}
		if strings.HasSuffix(hdr.Name, ".go") {
			t.Fatalf("include whitelist should drop .go files, got %q", hdr.Name)
		}
		if hdr.Name == "proj/docs/guide.md" {
			foundGuide = true
		}
		if hdr.Name == "proj/README.md" {
			foundReadme = true
		}
	}
	if !foundGuide {
		t.Fatal("guide.md missing")
	}
	if !foundReadme {
		t.Fatal("README.md missing")
	}
}

// TestArchiveDirIncludeNestedFileNotPrunedByParentDir reproduces the B2 bug:
// --include "src/app.go" must NOT prune the "src" directory (which itself does
// not match the include pattern), otherwise the leaf is never visited and
// the archive is empty. Pruning must be driven by MatchExcluded, not Match.
func TestArchiveDirIncludeNestedFileNotPrunedByParentDir(t *testing.T) {
	mock := newArchiveMockServer(map[string]string{
		"/proj/src/app.go":       "package main\n",
		"/proj/src/util/util.go": "package util\n",
		"/proj/other/notes.txt":  "notes\n",
		"/proj/README.md":        "# readme\n",
	})
	srv := mock.server(t)
	defer srv.Close()
	c := New(srv.URL, "")
	c.smallFileThreshold = DefaultSmallFileThreshold

	var buf bytes.Buffer
	if err := c.ArchiveDir(context.Background(), "/proj", &buf, ArchiveOptions{
		Include: []string{"src/app.go"},
	}); err != nil {
		t.Fatalf("ArchiveDir: %v", err)
	}
	gz, _ := gzip.NewReader(bytes.NewReader(buf.Bytes()))
	defer func() { _ = gz.Close() }()
	tr := tar.NewReader(gz)
	var names []string
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			t.Fatalf("tar next: %v", err)
		}
		names = append(names, hdr.Name)
	}
	if !containsStr(names, "proj/src/app.go") {
		t.Fatalf("include src/app.go missing (parent dir was pruned): %v", names)
	}
	// The other leaf must be dropped by the include whitelist.
	if containsStr(names, "proj/other/notes.txt") {
		t.Fatalf("notes.txt should be dropped by include whitelist: %v", names)
	}
	if containsStr(names, "proj/README.md") {
		t.Fatalf("README.md should be dropped by include whitelist: %v", names)
	}
}

func TestArchiveDirZipFormat(t *testing.T) {
	mock := newArchiveMockServer(map[string]string{
		"/proj/a.txt":   "AAA",
		"/proj/b/c.txt": "CCC",
	})
	srv := mock.server(t)
	defer srv.Close()
	c := New(srv.URL, "")
	c.smallFileThreshold = DefaultSmallFileThreshold

	var buf bytes.Buffer
	if err := c.ArchiveDir(context.Background(), "/proj", &buf, ArchiveOptions{Format: ArchiveFormatZip}); err != nil {
		t.Fatalf("ArchiveDir: %v", err)
	}
	zr, err := zip.NewReader(bytes.NewReader(buf.Bytes()), int64(buf.Len()))
	if err != nil {
		t.Fatalf("zip reader: %v", err)
	}
	var names []string
	for _, f := range zr.File {
		names = append(names, f.Name)
	}
	if !containsStr(names, "proj/a.txt") || !containsStr(names, "proj/b/c.txt") {
		t.Fatalf("zip entries = %v; want a.txt and b/c.txt", names)
	}
}

func TestArchiveDirFlatStripsHierarchy(t *testing.T) {
	mock := newArchiveMockServer(map[string]string{
		"/proj/src/deep/nested/a.go": "package nested\n",
		"/proj/b.go":                 "package main\n",
	})
	srv := mock.server(t)
	defer srv.Close()
	c := New(srv.URL, "")
	c.smallFileThreshold = DefaultSmallFileThreshold

	var buf bytes.Buffer
	if err := c.ArchiveDir(context.Background(), "/proj", &buf, ArchiveOptions{Flat: true}); err != nil {
		t.Fatalf("ArchiveDir: %v", err)
	}
	gz, _ := gzip.NewReader(bytes.NewReader(buf.Bytes()))
	defer func() { _ = gz.Close() }()
	tr := tar.NewReader(gz)
	var names []string
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			t.Fatalf("tar next: %v", err)
		}
		names = append(names, hdr.Name)
	}
	if !containsStr(names, "a.go") || !containsStr(names, "b.go") {
		t.Fatalf("flat entries = %v; want a.go and b.go", names)
	}
	for _, n := range names {
		if strings.Contains(n, "/") {
			t.Fatalf("flat mode should emit basenames only, got %q", n)
		}
	}
}

func TestArchiveDirFlatRejectsDuplicateBasenames(t *testing.T) {
	mock := newArchiveMockServer(map[string]string{
		"/proj/src/config.json":  "{}\n",
		"/proj/test/config.json": "{}\n",
	})
	srv := mock.server(t)
	defer srv.Close()
	c := New(srv.URL, "")
	c.smallFileThreshold = DefaultSmallFileThreshold

	var buf bytes.Buffer
	err := c.ArchiveDir(context.Background(), "/proj", &buf, ArchiveOptions{Flat: true})
	if err == nil {
		t.Fatal("expected collision error for duplicate basenames in flat mode")
	}
	if !strings.Contains(err.Error(), "collision") {
		t.Fatalf("expected collision error, got: %v", err)
	}
}

func TestArchiveDirDirectoryPruningSkipsExcludedSubtree(t *testing.T) {
	// Tree with a nested node_modules subtree. The mock records ListCtx calls
	// so we can assert that excluded subtrees are NOT listed.
	mock := newArchiveMockServer(map[string]string{
		"/proj/src/app.go":                    "package src\n",
		"/proj/node_modules/react/index.js":   "module.exports\n",
		"/proj/node_modules/react/foo/bar.js": "foo\n",
	})
	srv := mock.server(t)
	defer srv.Close()
	c := New(srv.URL, "")
	c.smallFileThreshold = DefaultSmallFileThreshold

	var buf bytes.Buffer
	if err := c.ArchiveDir(context.Background(), "/proj", &buf, ArchiveOptions{
		Exclude: []string{"**/node_modules/**"},
	}); err != nil {
		t.Fatalf("ArchiveDir: %v", err)
	}
	// The excluded subtree's nested directories must NOT have been listed —
	// directory pruning at BFS time means we never issue ListCtx for children
	// of an excluded directory.
	mock.mu.Lock()
	reactCalls := mock.listCalls["/proj/node_modules/react"]
	reactFooCalls := mock.listCalls["/proj/node_modules/react/foo"]
	mock.mu.Unlock()
	if reactCalls > 0 {
		t.Fatalf("pruned subtree /proj/node_modules/react was listed %d time(s); expected 0", reactCalls)
	}
	if reactFooCalls > 0 {
		t.Fatalf("pruned subtree /proj/node_modules/react/foo was listed %d time(s); expected 0", reactFooCalls)
	}
	// And the archive must contain no node_modules entries.
	gz, _ := gzip.NewReader(bytes.NewReader(buf.Bytes()))
	defer func() { _ = gz.Close() }()
	tr := tar.NewReader(gz)
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			t.Fatalf("tar next: %v", err)
		}
		if strings.Contains(hdr.Name, "node_modules") {
			t.Fatalf("excluded node_modules leaked into archive: %q", hdr.Name)
		}
	}
}

func TestArchiveDirInvalidFormat(t *testing.T) {
	c := New("http://127.0.0.1:1", "")
	var buf bytes.Buffer
	err := c.ArchiveDir(context.Background(), "/proj", &buf, ArchiveOptions{Format: ArchiveFormat("rar")})
	if err == nil {
		t.Fatal("expected error for invalid format")
	}
	if !strings.Contains(err.Error(), "tar.gz or zip") {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestArchiveDirInvalidPattern(t *testing.T) {
	c := New("http://127.0.0.1:1", "")
	var buf bytes.Buffer
	err := c.ArchiveDir(context.Background(), "/proj", &buf, ArchiveOptions{
		Exclude: []string{"**/../bad/**"},
	})
	if err == nil {
		t.Fatal("expected error for invalid pattern")
	}
}

func containsStr(s []string, v string) bool {
	for _, x := range s {
		if x == v {
			return true
		}
	}
	return false
}

File: clients/drive9-js/tests/archive.test.ts
import { describe, expect, it } from "vitest";
import { Client } from "../src/index.js";
import { setupServer } from "msw/node";
import { http, HttpResponse } from "msw";
import * as zlib from "node:zlib";

const server = setupServer();
server.listen({ onUnhandledRequest: "error" });

interface MockFile {
  path: string;
  body: string;
}

// Seed a mock remote tree from a flat map of {path: body}. Returns the per-dir
// listings to wire into the list endpoint.
function buildTree(files: MockFile[]): { dirs: Record<string, { name: string; size: number; isDir: boolean }[]>; bodies: Record<string, string> } {
  const dirs: Record<string, { name: string; size: number; isDir: boolean }[]> = {};
  const bodies: Record<string, string> = {};
  const dirSeen: Record<string, Set<string>> = {};
  for (const f of files) {
    bodies[f.path] = f.body;
    let dir = "/";
    const rel = f.path.replace(/^\/+/, "");
    const segs = rel.split("/");
    for (let i = 0; i < segs.length; i++) {
      const seg = segs[i];
      const isLast = i === segs.length - 1;
      if (!dirSeen[dir]) dirSeen[dir] = new Set();
      if (!dirSeen[dir].has(seg)) {
        dirSeen[dir].add(seg);
        dirs[dir] = dirs[dir] || [];
        dirs[dir].push({ name: seg, size: isLast ? f.body.length : 0, isDir: !isLast });
      }
      if (isLast) break;
      const child = dir === "/" ? `/${seg}` : `${dir}/${seg}`;
      if (!dirs[child]) dirs[child] = [];
      dir = child;
    }
  }
  return { dirs, bodies };
}

function mountTree(root: string, files: MockFile[]) {
  const { dirs, bodies } = buildTree(files.map((f) => ({ path: f.path, body: f.body })));
  server.use(
    http.get(`http://localhost:9009/v1/fs${root === "/" ? "" : root}`, ({ request }) => {
      const url = new URL(request.url);
      if (!url.searchParams.has("list")) {
        // GET file content
        const body = bodies[root];
        if (body === undefined) return HttpResponse.json({}, { status: 404 });
        return HttpResponse.text(body);
      }
      const entries = dirs[root] || [];
      return HttpResponse.json({ entries });
    }),
    http.get(/http:\/\/localhost:9009\/v1\/fs\/.+/, ({ request }) => {
      const url = new URL(request.url);
      const p = url.pathname.replace("/v1/fs", "");
      if (url.searchParams.has("list")) {
        const entries = dirs[p] || [];
        return HttpResponse.json({ entries });
      }
      const body = bodies[p];
      if (body === undefined) return HttpResponse.json({}, { status: 404 });
      return HttpResponse.text(body);
    })
  );
}

// Minimal ustar reader: returns entry names (dirs get trailing slash).
function readTarGz(buf: Buffer): string[] {
  const decompressed = zlib.gunzipSync(buf);
  const names: string[] = [];
  let offset = 0;
  while (offset + 512 <= decompressed.length) {
    const header = decompressed.subarray(offset, offset + 512);
    // All-zero block → EOF.
    let allZero = true;
    for (let i = 0; i < 512; i++) {
      if (header[i] !== 0) { allZero = false; break; }
    }
    if (allZero) break;
    const name = header.subarray(0, 100).toString("utf8").replace(/\0+$/, "");
    const typeflag = String.fromCharCode(header[156]);
    const sizeStr = header.subarray(124, 136).toString("utf8").replace(/\0+$/, "").trim();
    const size = parseInt(sizeStr, 8) || 0;
    const entryName = typeflag === "5" ? name.replace(/\/?$/, "/") : name;
    names.push(entryName);
    offset += 512 + Math.ceil(size / 512) * 512;
  }
  return names.sort();
}

async function streamToBuffer(stream: ReadableStream<Uint8Array>): Promise<Buffer> {
  const reader = stream.getReader();
  const chunks: Buffer[] = [];
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    chunks.push(Buffer.from(value));
  }
  return Buffer.concat(chunks);
}

describe("archive", () => {
  it("produces a tar.gz with the remote tree", async () => {
    mountTree("/proj", [
      { path: "/proj/README.md", body: "hello world\n" },
      { path: "/proj/src/main.go", body: "package main\n" },
      { path: "/proj/src/util/util.go", body: "package util\n" },
    ]);
    const client = new Client("http://localhost:9009", "test-key");
    const stream = await client.archive("/proj");
    const buf = await streamToBuffer(stream);
    const names = readTarGz(buf);
    expect(names).toContain("proj/");
    expect(names).toContain("proj/README.md");
    expect(names).toContain("proj/src/main.go");
    expect(names).toContain("proj/src/util/util.go");
  });

  it("excludes node_modules via --exclude", async () => {
    mountTree("/proj", [
      { path: "/proj/src/app.go", body: "package src\n" },
      { path: "/proj/node_modules/react/index.js", body: "module.exports\n" },
    ]);
    const client = new Client("http://localhost:9009", "test-key");
    const stream = await client.archive("/proj", { exclude: ["**/node_modules/**"] });
    const names = readTarGz(await streamToBuffer(stream));
    for (const n of names) {
      expect(n.includes("node_modules")).toBe(false);
    }
    expect(names).toContain("proj/src/app.go");
  });

  it("coding-agent profile skips default excludes", async () => {
    mountTree("/proj", [
      { path: "/proj/main.go", body: "package main\n" },
      { path: "/proj/dist/bundle.js", body: "bundle\n" },
      { path: "/proj/node_modules/react/x.js", body: "x\n" },
      { path: "/proj/.git/HEAD", body: "ref: main\n" },
    ]);
    const client = new Client("http://localhost:9009", "test-key");
    const stream = await client.archive("/proj", { profile: "coding-agent" });
    const names = readTarGz(await streamToBuffer(stream));
    expect(names).toContain("proj/main.go");
    for (const n of names) {
      expect(n.includes("node_modules") || n.includes(".git/") || n.includes("dist/")).toBe(false);
    }
  });

  it("include whitelist keeps only matching paths", async () => {
    mountTree("/proj", [
      { path: "/proj/src/app.go", body: "package src\n" },
      { path: "/proj/docs/guide.md", body: "# guide\n" },
      { path: "/proj/README.md", body: "# readme\n" },
    ]);
    const client = new Client("http://localhost:9009", "test-key");
    const stream = await client.archive("/proj", { include: ["docs/**", "README.md"] });
    const names = readTarGz(await streamToBuffer(stream));
    expect(names).toContain("proj/docs/guide.md");
    expect(names).toContain("proj/README.md");
    for (const n of names) {
      expect(n.endsWith(".go")).toBe(false);
    }
  });

  it("flat mode strips hierarchy", async () => {
    mountTree("/proj", [
      { path: "/proj/src/deep/nested/a.go", body: "package nested\n" },
      { path: "/proj/b.go", body: "package main\n" },
    ]);
    const client = new Client("http://localhost:9009", "test-key");
    const stream = await client.archive("/proj", { flat: true });
    const names = readTarGz(await streamToBuffer(stream));
    for (const n of names) {
      expect(n.endsWith("/")).toBe(false); // no dir entries
      expect(n.includes("/")).toBe(false); // basenames only
    }
    expect(names).toContain("a.go");
    expect(names).toContain("b.go");
  });

  it("zip format via archiveToFile", async () => {
    mountTree("/proj", [
      { path: "/proj/a.txt", body: "AAA" },
      { path: "/proj/b/c.txt", body: "CCC" },
    ]);
    const client = new Client("http://localhost:9009", "test-key");
    const out = `/tmp/drive9-ts-zip-${Date.now()}.zip`;
    await client.archiveToFile("/proj", out, { format: "zip" });
    const fs = await import("node:fs");
    const data = fs.readFileSync(out);
    // Validate the zip is well-formed and has the expected entries.
    const { execSync } = await import("node:child_process");
    const list = execSync(`python3 -c 'import zipfile,sys; print("\\n".join(sorted(zipfile.ZipFile(sys.argv[1]).namelist())))' ${out}`, { encoding: "utf8" });
    expect(list).toContain("proj/a.txt");
    expect(list).toContain("proj/b/c.txt");
    // testzip returns None when valid.
    const valid = execSync(`python3 -c 'import zipfile,sys; print("valid" if zipfile.ZipFile(sys.argv[1]).testzip() is None else "corrupt")' ${out}`, { encoding: "utf8" }).trim();
    expect(valid).toBe("valid");
    fs.unlinkSync(out);
  });

  it("flat mode rejects duplicate basenames", async () => {
    mountTree("/proj", [
      { path: "/proj/src/config.json", body: "{}\n" },
      { path: "/proj/test/config.json", body: "{}\n" },
    ]);
    const client = new Client("http://localhost:9009", "test-key");
    await expect(client.archive("/proj", { flat: true })).rejects.toThrow(/collision/);
  });

  it("include nested file does not prune parent directory", async () => {
    // Reproduces B2: --include "src/app.go" must walk into src/ even though
    // "src" itself does not match the include pattern, otherwise the leaf is
    // never visited and the archive is empty.
    mountTree("/proj", [
      { path: "/proj/src/app.go", body: "package main\n" },
      { path: "/proj/src/util/util.go", body: "package util\n" },
      { path: "/proj/other/notes.txt", body: "notes\n" },
    ]);
    const client = new Client("http://localhost:9009", "test-key");
    const stream = await client.archive("/proj", { include: ["src/app.go"] });
    const names = readTarGz(await streamToBuffer(stream));
    expect(names).toContain("proj/src/app.go");
    expect(names.some((n) => n.endsWith("notes.txt"))).toBe(false);
  });

  it("long entry names use a PAX header instead of truncating", async () => {
    // A path > 100 bytes must survive via a PAX 'x' extended header. Use
    // python's tarfile to parse since it honors PAX records (the minimal
    // ustar reader in this test file does not).
    const longDir =
      "/proj/very/deeply/nested/path/that/definitely/exceeds/the/ustar/one/hundred/byte/name/field/limit";
    const longFile = `${longDir}/module.ts`;
    mountTree("/proj", [{ path: longFile, body: "export {}\n" }]);
    const client = new Client("http://localhost:9009", "test-key");
    const stream = await client.archive("/proj");
    const buf = await streamToBuffer(stream);
    const fs = await import("node:fs");
    const tmp = `/tmp/drive9-ts-longname-${Date.now()}.tar.gz`;
    fs.writeFileSync(tmp, buf);
    try {
      const { execSync } = await import("node:child_process");
      const list = execSync(
        `python3 -c 'import tarfile,sys; [print(m.name) for m in tarfile.open(sys.argv[1],"r:gz")]' ${tmp}`,
        { encoding: "utf8" }
      );
      // The full long path must appear (archiveName prepends the archive root
      // "proj/", and longFile already starts with "/proj/...", so the in-archive
      // name is longFile with its leading "/" stripped).
      const expectedName = longFile.slice(1);
      expect(list).toContain(expectedName);
      // Sanity: the name is genuinely > 100 bytes so the PAX path was needed.
      expect(expectedName.length).toBeGreaterThan(100);
    } finally {
      fs.unlinkSync(tmp);
    }
  });
});